File: remocra/src/main/webapp/js/app/remocra/model/Permis.js
Ext.require('Sdis.Remocra.model.Commune');
Ext.require('Sdis.Remocra.model.TypeOrganisme');
Ext.require('Sdis.Remocra.model.TypePermisAvis');
Ext.require('Sdis.Remocra.model.TypePermisInterservice');
Ext.require('Sdis.Remocra.model.Utilisateur');
Ext.require('Sdis.Remocra.model.PermisDocument');

Ext.define('Sdis.Remocra.model.Permis', {
    extend : 'Ext.data.Model',

    // Champs supplémentaires
    fields : [
            {
                name : 'id',
                type : 'int',
                useNull : true
            },
            {
                name : 'dateModification',
                type : 'date',
                dateFormat : 'c'
            },
            {
                name : 'datePermis',
                type : 'date',
                dateFormat : 'c'
            },
            {
                name : 'nom',
                type : 'string'
            },
            {
                name : 'voie',
                type : 'string'
            },
            {
                name : 'complement',
                type : 'string'
            },
            {
                name : 'numero',
                type : 'string'
            },
            {
                name : 'sectionCadastrale',
                type : 'string'
            },
            {
                name : 'parcelleCadastrale',
                type : 'string'
            },
            {
                name : 'observations',
                type : 'string'
            },
            {
                name : 'annee',
                type : 'int'
            },
            {
                name : 'geometrie',
                type : 'string'
            },
            {
                name : 'display',
                persist : false,
                convert : function(v, record) {
                    return record.get('annee') + ' - ' + record.get('numero')
                            + ' ' + record.get('nom');
                }
            } ],

    // Validations supplémentaires
    validations : [],

    associations : [ {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Commune',
        associationKey : 'commune',
        getterName : 'getCommune',
        setterName : 'setCommune',
        associatedName : 'Commune',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypePermisAvis',
        associationKey : 'avis',
        getterName : 'getAvis',
        setterName : 'setAvis',
        associatedName : 'Avis',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypePermisInterservice',
        associationKey : 'interservice',
        getterName : 'getInterservice',
        setterName : 'setInterservice',
        associatedName : 'Interservice',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.TypeOrganisme',
        associationKey : 'serviceInstructeur',
        getterName : 'getServiceInstructeur',
        setterName : 'setServiceInstructeur',
        associatedName : 'ServiceInstructeur',
        persist : true
    }, {
        type : 'belongsTo',
        model : 'Sdis.Remocra.model.Utilisateur',
        associationKey : 'instructeur',
        getterName : 'getInstructeur',
        setterName : 'setInstructeur',
        associatedName : 'Instructeur',
        persist : false
    }, {
        type : 'hasMany',
        model : 'Sdis.Remocra.model.PermisDocument',
        associationKey : 'permisDocuments',
        name : 'permisDocuments',
        associatedName : 'permisDocuments',
        persist : false
    } ]
});

File: remocra/src/main/java/fr/sdis83/remocra/domain/remocra/Permis.java
package fr.sdis83.remocra.domain.remocra;

import java.util.Date;
import java.util.HashSet;
import java.util.Set;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.persistence.Version;
import javax.validation.constraints.NotNull;

import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.hibernate.annotations.Type;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.roo.addon.javabean.RooJavaBean;
import org.springframework.roo.addon.jpa.activerecord.RooJpaActiveRecord;
import org.springframework.roo.addon.json.RooJson;
import org.springframework.roo.addon.tostring.RooToString;

import com.vividsolutions.jts.geom.Geometry;
import com.vividsolutions.jts.geom.Point;

import flexjson.JSONDeserializer;
import flexjson.JSONSerializer;
import fr.sdis83.remocra.domain.utils.RemocraDateHourTransformer;
import fr.sdis83.remocra.web.deserialize.GeometryFactory;

@RooJavaBean
@RooToString
@RooJson
@RooJpaActiveRecord(versionField = "version")
public class Permis {

    static JSONSerializer SERIALIZER = new JSONSerializer();

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Version
    @Column(name = "version", columnDefinition = "INTEGER default 1")
    private Integer version;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
    private Date dateModification;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @DateTimeFormat(pattern = RemocraDateHourTransformer.FORMAT)
    private Date datePermis;

    @NotNull
    private String nom;

    @NotNull
    @ManyToOne
    private Commune commune;

    @NotNull
    private String voie;

    @NotNull
    private String complement;

    @NotNull
    private String numero;

    @NotNull
    @ManyToOne
    private TypePermisAvis avis;

    @NotNull
    @ManyToOne
    private TypePermisInterservice interservice;

    @NotNull
    private String sectionCadastrale;

    @NotNull
    private String parcelleCadastrale;

    @NotNull
    private String observations;

    @NotNull
    @ManyToOne
    private TypeOrganisme serviceInstructeur;

    @NotNull
    private Integer annee;

    @NotNull
    @ManyToOne
    private Utilisateur instructeur;

    @NotNull
    @Type(type = "org.hibernate.spatial.GeometryType")
    private Point geometrie;

    @OnDelete(action = OnDeleteAction.CASCADE)
    @OneToMany(cascade = CascadeType.ALL, mappedBy = "permis", orphanRemoval = true)
    private Set<PermisDocument> permisDocuments = new HashSet<PermisDocument>();

    public static Permis fromJsonToPermis(String json) {
        return new JSONDeserializer<Permis>().use(null, Permis.class).use(Date.class, RemocraDateHourTransformer.getInstance()).use(Geometry.class, new GeometryFactory())
                .deserialize(json);
    }

}
